main/src/ca/uwaterloo/flix/language/ast/Ast.scala

```diff
@@ -385,18 +392,27 @@ object Ast {
         ConstraintGraph(this.xs ++ that.xs)
     }
 
+    /**
+      * Returns true if the two types are relations or lattices and have the same arity.
+      */
+    private def typesMatch(t1: Type, t2: Type): Boolean =
```

Comment:
How about somethhing like `mayEqual` with a description that it returns false if it can definitely be ruled out that the two types are equal.